File: 056. Merge Intervals/C++/Solution.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<Interval> merge(vector<Interval> &intervals) {
        if (intervals.empty()) {
            return vector<Interval>();
        }
        sort(intervals.begin(), intervals.end(), [](const Interval & a, const Interval & b) {
            return a.start < b.start;
        });
        int start = intervals[0].start;
        int end = intervals[0].end;
        vector<Interval> result;
        for (const auto &interval : intervals) {
            if (interval.start > end) {
                result.push_back(Interval(start, end));
                start = interval.start;
            }
            end = max(end, interval.end);
        }
        result.push_back(Interval(start, end));
        return result;
    }
};

File: 056. Merge Intervals/Java/Solution.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
public class Solution {
    public List<Interval> merge(List<Interval> intervals) {
        if (intervals.isEmpty()) {
            return new ArrayList<>();
        }
        intervals.sort(new Comparator<Interval>() {
            @Override
            public int compare(Interval o1, Interval o2) {
                return o1.start - o2.start;
            }
        });
        int start = intervals.get(0).start;
        int end = intervals.get(0).end;
        List<Interval> result = new ArrayList<>();
        for (Interval interval : intervals) {
            if (interval.start > end) {
                result.add(new Interval(start, end));
                start = interval.start;
            }
            end = Math.max(end, interval.end);
        }
        result.add(new Interval(start, end));
        return result;
    }
}